api/server/server.go

```diff
@@ -536,6 +545,49 @@ func WithTracer(zipkinURL string) ServerOption {
 	}
 }
 
+// TODO plumbing considerations, we've put the S pipe next to the chandalier...
+func dockerPid() (int, error) {
+	// prometheus' process collector only works on linux anyway. let them do the
+	// process detection, if we return an error here we just get 0 metrics and it
+	// does not log / blow up (that's fine!) it's also likely we hit permissions
+	// errors here for many installations, we want to do similar and ignore (we
+	// just want for prod).
+
+	var pid int
+	err := filepath.Walk("/proc", func(path string, info os.FileInfo, err error) error {
```

Comment:
full walk seems expensive, but for now OK, we should improve this with list [\+d] dirs at first level, then read /proc/<pid>/status for each. (There's also rumours of truncation in Name: <s> if larger than 15 chars, so we might want to do a readlink '/proc/<pid>/exe' instead...) For future PRs...  I'm thinking we have an env var that has list of processes: ['dockerd', 'containerd', 'fn', ...]

Reply:
it is most definitely not ideal. dockerd can die and get a new pid, so this handles that. if we list from env it could go stale if dockerd restarts :/

the process should probably always be called `dockerd` and we won't hit name conflicts, but i know certain people like to make things more difficult than they need to be too. in any event, not catastrophic.

> we should improve this with list [+d] dirs at first level, then read /proc//status for each.

not following here, sorry. open to ideas on optimizing filepath walk. could probably cache the pid until it stops working, too.

interesting idea to track `containerd`, too